models/template_helpers.go

```diff
@@ -0,0 +1,222 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package models
+
+import (
+	"time"
+
+	"github.com/percona-platform/saas/pkg/alert"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+)
+
+func checkUniqueTemplateName(q *reform.Querier, name string) error {
+	if name == "" {
+		panic("empty template name")
+	}
+
+	template := &Template{Name: name}
+	switch err := q.Reload(template); err {
+	case nil:
+		return status.Errorf(codes.AlreadyExists, "Template with name %q already exists.", name)
+	case reform.ErrNoRows:
+		return nil
+	default:
+		return errors.WithStack(err)
+	}
+}
+
+// FindTemplates returns saved notification rule templates.
+func FindTemplates(q *reform.Querier) ([]Template, error) {
+	structs, err := q.SelectAllFrom(TemplateTable, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to select notification rule templates")
+	}
+
+	templates := make([]Template, len(structs))
+	for i, s := range structs {
+		c := s.(*Template)
+
+		templates[i] = *c
+	}
+
+	return templates, nil
+}
+
+// FindTemplateByName finds template by name.
+func FindTemplateByName(q *reform.Querier, name string) (*Template, error) {
```

Comment:
this function is same as the `checkUniqueTemplateName` except for the panic, why don't we drop the panic and reuse the code?

Reply:
Idea is that invocation `checkUniqueTemplateName` with empty rule name is the developer's mistake. For example, for agents, we create IDs on the server side, so they always should be present. In templates name provided by user, however, template `should` have a name, so we check it with `rule.Validate()` and notify user about bad request. So if `checkUniqueTemplateName` invoked with an empty name it's not something that can be handled, it's just serve bug.
On the other side, we have `FindTemplateByName` and here we have a different situation. Name for query provided by the user, so we should just notify him about the wrong request.

Reply:
Also `checkUniqueTemplateName` and `FindTemplateByName` have different errors/switch logic.

Reply:
got it, thanks